service/worker/workerdeployment/workflow.go

```diff
@@ -116,6 +127,105 @@ func (d *WorkflowRunner) run(ctx workflow.Context) error {
 	return workflow.NewContinueAsNewError(ctx, Workflow, d.WorkerDeploymentWorkflowArgs)
 }
 
+func (d *WorkflowRunner) validateSetWorkerDeploymentRampingVersion(args *deploymentspb.SetWorkerDeploymentRampingVersionArgs) error {
+	if args.Version == d.State.RoutingInfo.RampingVersion && args.Percentage == d.State.RoutingInfo.RampingVersionPercentage {
+		d.logger.Info("version already ramping, no change")
+		return temporal.NewApplicationError("version already ramping, no change", errNoChangeType)
+	}
+	// todo: this will only work when "__unversioned__" is the default current-version of a deployment.
+	if args.Version == d.State.RoutingInfo.CurrentVersion {
+		d.logger.Info("version can't be set to ramping since it is already current")
+		return temporal.NewApplicationError("version can't be set to ramping since it is already current", errVersionAlreadyCurrentType)
+	}
+
+	return nil
+}
+
+func (d *WorkflowRunner) handleSetWorkerDeploymentRampingVersion(ctx workflow.Context, args *deploymentspb.SetWorkerDeploymentRampingVersionArgs) (*deploymentspb.SetWorkerDeploymentRampingVersionResponse, error) {
+	// use lock to enforce only one update at a time
+	err := d.lock.Lock(ctx)
+	if err != nil {
+		d.logger.Error("Could not acquire workflow lock")
+		return nil, serviceerror.NewDeadlineExceeded("Could not acquire workflow lock")
+	}
+	d.pendingUpdates++
+	defer func() {
+		d.pendingUpdates--
+		d.lock.Unlock()
+	}()
+
+	prevRampingVersion := d.State.RoutingInfo.RampingVersion
+	prevRampingVersionPercentage := d.State.RoutingInfo.RampingVersionPercentage
+
+	newRampingVersion := args.Version
+	routingUpdateTime := timestamppb.New(workflow.Now(ctx))
+
+	var rampingSinceTime *timestamppb.Timestamp
+
+	// unsetting ramp
+	if newRampingVersion == "" {
+		if prevRampingVersion == "" {
+			// no previously set ramping version
+			return &deploymentspb.SetWorkerDeploymentRampingVersionResponse{}, nil
+		}
+
+		rampUpdateArgs := &deploymentspb.SyncVersionStateUpdateArgs{
+			RoutingUpdateTime: routingUpdateTime,
+			RampingSinceTime:  nil, // remove ramp
+			RampPercentage:    0,   // remove ramp
+		}
+
+		if _, err := d.syncVersion(ctx, prevRampingVersion, rampUpdateArgs); err != nil {
+			return nil, err
+		}
```

Comment:
Nit: You could put all of these `d.syncVersion(ctx, prevRampingVersion, rampUpdateArgs)` calls outside of the if-else block, and just use the if-else block to construct the appropriate `rampUpdateArgs`